py/usageaccountant/datadog_fetcher.py

```diff
@@ -0,0 +1,292 @@
+from typing import Any, Dict, List, Optional, Tuple, TypedDict
+from urllib.request import Request, urlopen
+
+from usageaccountant.accumulator import (
+    KafkaConfig,
+    UsageAccumulator,
+    UsageUnit,
+)
+
+# TODO fetch these from env
+headers = {"DD-APPLICATION-KEY": "", "DD-API-KEY": ""}
+
+logger = logging.getLogger("fetcher")
+
+# TODO remove this before prod
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s [%(levelname)s] %(message)s",
+    handlers=[logging.FileHandler("debug.log"), logging.StreamHandler()],
+)
+
+
+class DatadogResponseUnit(TypedDict):
+    plural: str
+    name: str
+
+
+class DatadogResponseSeries(DatadogResponseUnit):
+    unit: List[DatadogResponseUnit]
+    pointlist: List[List[float]]
+    scope: str
+
+
+class DatadogResponse(DatadogResponseSeries):
+    error: Optional[str]
+    query: str
+    from_date: int
+    to_date: int
+    status: str
+    series: List[DatadogResponseSeries]
+
+
+def is_valid_query(query: str) -> bool:
+    """
+    Validates if query string contains shared_resource_id
+    and app_feature–parameters used in UsageAccountant.record().
+
+    query: Datadog query
+    """
+    return "shared_resource_id" in query and "app_feature" in query
+
+
+def get(query: str, start_time: int, end_time: int) -> Any:
+    """
+    Fetches timeseries data from Datadog API, returns response dict.
+
+    query: Datadog query
+    start_time: Start of the time window in Unix epoch format
+                with second-level precision
+    end_time: End of the time window in Unix epoch format
+              with second-level precision
+    """
+    data = {"query": query, "from": start_time, "to": end_time}
+
+    base_url = "https://api.datadoghq.com/api/v1/query"
+    query_string = urllib.parse.urlencode(data)
+    full_url = f"{base_url}?{query_string}"
+
+    try:
+        request = Request(url=full_url, headers=headers)
+        # context is needed to handle
+        # ssl.SSLCertVerificationError:
+        # [SSL: CERTIFICATE_VERIFY_FAILED] on macOS
+        # TODO remove next three lines and
+        # context parameter from the call to urlopen()
+        import ssl
+
+        context = ssl.SSLContext()
+        response = urlopen(request, context=context).read()
+        return json.loads(response)
+
+    except Exception as e:
+        raise e
+
+
+def parse_response_series_and_unit(
+    response: DatadogResponse,
+) -> Tuple[List[DatadogResponseSeries], UsageUnit]:
+    """
+    Extracts list of series objects and unit from the API response.
+
+    response: response received from API call
+    """
+
+    error_msg = response.get("error")
+    if error_msg:
+        raise Exception(format_exception_message(response, error_msg))
+
+    if response.get("series") is not None:
+        series_list = response.get("series", [])
+        if len(series_list) > 0:
+            first_series = series_list[0]
+            if first_series.get("unit") is not None:
```

Comment:
nit:
when you have a series of conditions like here and you do something only if they all pass, consider inverting them and returning earlier.

```
if not len(series_list):
  return
if not first_series.get(...):
  return
....
do stuff.
```

Reply:
This pattern is entirely designed to avoid too many indentation, if you still layout the code the same way and rely on checks at the end to raise exception you do not gain anything. Do the check that rasies the exception first and raise right away without indenting.